### PR TITLE
refactor internal codes (to_kana)

### PR DESCRIPTION
**Pull request by JSYoo5B**
On resolving bugfix about InputContext, I'm doubting `to_kana()` method.

In order to fix bugs in InputContext, I think refactoring is necessary for debugging.
No offence, but it's difficult for me to understand code flow with the current codebase.

I'm just trying to refactor internal codes, not the public interfaces & constants.
Detailed reasons for its refactoring were described in each commit log.
I suppose this refactor may not finish very soon. So I'm marking it with WIP

### CODE REVIEW

**Comment by onting** [comment]
Your refactoring seems going well so far.

**Comment by onting** [comment]
I guess _IDX infix in kana idxs can be omitted.
I feel we can know that the enums are table idxs clearly without the infix.
How do you think about it?
@JSYoo5B

**Comment by JSYoo5B** [comment]
Well, I added infix `IDX` to mark the difference between the Unicode constants and indexing constants.
Maybe this idea was derived from `HANGUL` Unicode constants naming. (Hangul has each constant for each Jaso)

Anyway, your opinion is reasonable. I'll consider it. (and maybe apply)
Thx.

**Comment by onting** [comment]
@JSYoo5B Can you give me a guarantee when this job would be finished?
#39 would goes with some code changes, and I want do it on your refactoring.
Thanks.

**Comment by JSYoo5B** [comment]
@onting Sorry for the lazy updates.
I'll try to finish this refactor in the upcoming Sunday.

**Comment by onting** [comment]
Good, #44 meight affect your affect.

**Comment by JSYoo5B** [comment]
Rebased current refactoring into a recent bug-fixed version.

**Comment by JSYoo5B** [comment]
@onting I extracted `HanjpBuffer` from `HanjpAutomata`

They're strongly coupled. Most of the code blocks have their purpose. But they aren't abstracted enough. I'm trying to abstract them into more readable functions.

How do you think about these commits?

**Comment by JSYoo5B** [comment]
The next remaining tasks are following:

1. Consume(eat) Choseong and Jungseong into Kana indexing => to internal functions
2. Reduce and divide combined Jungseong => to internal functions
3. Rename some short named variables (ex. `r`, `c`) and reused variables into proper name.

**Comment by onting** [comment]
> @onting I extracted `HanjpBuffer` from `HanjpAutomata`
> 
> They're strongly coupled. Most of the code blocks have their purpose. But they aren't abstracted enough. I'm trying to abstract them into more readable functions.
> 
> How do you think about these commits?

I think it is good.
I had little worry about it when I first saw it because I wanted hide all buffer class operation except instanciation.
However, it can be safe with some protection.
So It feels okay.

**Comment by onting** [comment]
> The next remaining tasks are following:
> 
> 1. Consume(eat) Choseong and Jungseong into Kana indexing => to internal functions
> 2. Reduce and divide combined Jungseong => to internal functions
> 3. Rename some short named variables (ex. `r`, `c`) and reused variables into proper name.

I think '2' is unnecessary.
I don't want add many hangul manipulating function explicitly.
Most of them are defined in libhangul, and I think additional operations are simple enough.
I think '3' is good.

**Comment by JSYoo5B** [comment]
@onting I finished refactoring related with `to_kana()` methods.

There are some other opinions about refactoring, but I'll write it as an issue. (This opinion may against your thought, so it could be a long discussion. Also, we need to prepare for releasing into stable version.)

File: hanjpautomata.c [hunk 1]
```diff
@@ -120,7 +44,8 @@ static const gunichar kana_table[][5] = {
     {0x3089, 0x308A, 0x308B, 0x308C, 0x308D}, // RA
     {0x308F, 0x3090, 0x0000, 0x3091, 0x3092}  // WA
 };
-static const gunichar kana_nn = 0x3093;
+static const gunichar KANA_SMALL_TU = 0x3063;
```
Comment (onting) [inline hanjpautomata.c:47]: Can you change it to KANA_SMALL_TSU?

**Review by onting — COMMENTED** [review]
Your work seems good except bad indentation.
But it meight be fixed with some automated way.

File: hanjpautomata.c (hunk 1 above)
Comment (JSYoo5B) [reply hanjpautomata.c:47]: I also think `KANA_SMALL_TSU` is better. But, according to the Unicode, it has been defined as `SMALL_TU`. see following link: https://unicode-table.com/en/3063/

Comment (onting) [reply hanjpautomata.c:47]: Alright.
Then, I also think tu is better.

**Review by onting — APPROVED** [review]

**Pull request merged**